<h3><span class=" glyphicon glyphicon-tag"/>&nbspExecutionTester</h3><kbd>Executes methods or other tests from the same test unit</kbd><br><br><br><h4 style="margin:0px">Class: org.apache.commons.exec.DefaultExecutorTest </h4><pre class="type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * Invoke the error script but define that the ERROR_STATUS is a good
 * exit value and therefore no exception should be thrown.
 * @throws Exception the test failed
 */
@Test public void testExecuteWithCustomExitValue1() throws Exception {
  exec.setExitValue(ERROR_STATUS);
  final CommandLine cl=new CommandLine(errorTestScript);
  exec.execute(cl);
}

</code></pre>

<br>
<pre class="type-13 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Executes methods or other tests from the same test unit
"></span><br>
@Test(expected=IOException.class) public void testExecuteWithInvalidWorkingDirectory() throws Exception {
  final File workingDir=new File("/foo/bar");
  final CommandLine cl=new CommandLine(testScript);
  exec.setWorkingDirectory(workingDir);
  exec.execute(cl);
}

</code></pre>

<br>
<pre class="type-13 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * Try to start an non-existing application which should result
 * in an exception.
 */
@Test(expected=IOException.class) public void testExecuteNonExistingApplicationWithWatchDog() throws Exception {
  final CommandLine cl=new CommandLine(nonExistingTestScript);
  final DefaultExecutor executor=new DefaultExecutor();
  executor.setWatchdog(new ExecuteWatchdog(ExecuteWatchdog.INFINITE_TIMEOUT));
  executor.execute(cl);
}

</code></pre>

<br>
<pre class="type-13 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * Try to start an non-existing application which should result
 * in an exception.
 */
@Test(expected=IOException.class) public void testExecuteNonExistingApplication() throws Exception {
  final CommandLine cl=new CommandLine(nonExistingTestScript);
  final DefaultExecutor executor=new DefaultExecutor();
  executor.execute(cl);
}

</code></pre>

<br>
<h4 style="margin:0px">Class: org.apache.commons.exec.issues.Exec49Test </h4><pre class="type-15 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Invokes logging operations">Logger</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Invokes logging operations
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * The issue was detected when trying to capture stdout/stderr with a PipedOutputStream and
 * then pass that to a PipedInputStream. The following code will produce the error.
 * The reason for the error is the PipedOutputStream is not being closed correctly,
 * causing the PipedInputStream to break.
 * @throws Exception the test failed
 */
@Test public void testExec49_1() throws Exception {
  if (OS.isFamilyUnix()) {
    final CommandLine cl=CommandLine.parse("/bin/ls");
    cl.addArgument("/opt");
    final PipedOutputStream pipedOutputStream=new PipedOutputStream();
    final PumpStreamHandler psh=new PumpStreamHandler(pipedOutputStream);
    exec.setStreamHandler(psh);
    System.out.println("Preparing to execute process - commandLine=" + cl.toString());
    final DefaultExecuteResultHandler handler=new DefaultExecuteResultHandler();
    exec.execute(cl,handler);
    System.out.println("Process spun off successfully - process=" + cl.getExecutable());
    int x;
    final PipedInputStream pis=new PipedInputStream(pipedOutputStream);
    while ((x=pis.read()) >= 0) {
    }
    pis.close();
    handler.waitFor(10000);
    handler.getExitValue();
  }
}

</code></pre>

<br>
<pre class="type-15 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Invokes logging operations">Logger</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Invokes logging operations
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * The issue was detected when trying to capture stdout with a PipedOutputStream and
 * then pass that to a PipedInputStream. The following code will produce the error.
 * The reason for the error is the PipedOutputStream is not being closed correctly,
 * causing the PipedInputStream to break.
 * @throws Exception the test failed
 */
@Test public void testExec49_2() throws Exception {
  if (OS.isFamilyUnix()) {
    final CommandLine cl=CommandLine.parse("/bin/ls");
    cl.addArgument("/opt");
    final PipedOutputStream pipedOutputStream=new PipedOutputStream();
    final PumpStreamHandler psh=new PumpStreamHandler(pipedOutputStream,new ByteArrayOutputStream());
    exec.setStreamHandler(psh);
    System.out.println("Preparing to execute process - commandLine=" + cl.toString());
    final DefaultExecuteResultHandler handler=new DefaultExecuteResultHandler();
    exec.execute(cl,handler);
    System.out.println("Process spun off successfully - process=" + cl.getExecutable());
    int x;
    final PipedInputStream pis=new PipedInputStream(pipedOutputStream);
    while ((x=pis.read()) >= 0) {
    }
    pis.close();
    handler.waitFor(10000);
    handler.getExitValue();
  }
}

</code></pre>

<br>
<h4 style="margin:0px">Class: org.apache.commons.exec.issues.Exec57Test </h4><pre class="type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * The same approach using a completely detached process works nicely on Mac OS X.
 * @throws IOException
 */
@Test(timeout=TEST_TIMEOUT) public void testExecutionOfDetachedProcess() throws IOException {
  if (!OS.isFamilyUnix()) {
    testNotSupportedForCurrentOperatingSystem();
    return;
  }
  final CommandLine cmdLine=new CommandLine("sh").addArgument("-c").addArgument("./src/test/scripts/issues/exec-57-detached.sh",false);
  final DefaultExecutor executor=new DefaultExecutor();
  final PumpStreamHandler pumpStreamHandler=new PumpStreamHandler(System.out,System.err);
  executor.setStreamHandler(pumpStreamHandler);
  executor.execute(cmdLine);
}

</code></pre>

<br>
<pre class="type-12 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Is not executed with the test suite">IgnoredMethod</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Is not executed with the test suite
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * DefaultExecutor.execute() does not return even if child process terminated - in this
 * case the child process hangs because the grand children is connected to stdout & stderr
 * and is still running. As work-around a stop timeout is used for the PumpStreamHandler
 * to ensure that the caller does not block forever but if the stop timeout is exceeded
 * an ExecuteException is thrown to notify the caller. But this case the threads are still
 * around causing a resource leak.
 * @TODO [EXEC-57] Broken for Mac OS X & Linux
 */
@Ignore("Broken for Unix-based systems") @Test(timeout=TEST_TIMEOUT) public void testExecutionOfBackgroundProcess() throws IOException {
  final CommandLine cmdLine=new CommandLine("sh").addArgument("-c").addArgument("./src/test/scripts/issues/exec-57-nohup.sh",false);
  final DefaultExecutor executor=new DefaultExecutor();
  final PumpStreamHandler pumpStreamHandler=new PumpStreamHandler(System.out,System.err);
  executor.setStreamHandler(pumpStreamHandler);
  executor.execute(cmdLine);
}

</code></pre>

<br>
<h4 style="margin:0px">Class: org.apache.commons.exec.issues.Exec62Test </h4><pre class="type-13 type-12 type-6 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Is not executed with the test suite">IgnoredMethod</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Contains more than 2 JUnit-based stereotypes">HybridVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Is not executed with the test suite
- Contains more than 2 JUnit-based stereotypes
- Executes methods or other tests from the same test unit
"></span><br>
@Ignore("Test behaves differently between Mac OS X and Linux - don't know why") @Test(expected=TimeoutException.class,timeout=10000) public void testMe() throws Exception {
  if (OS.isFamilyUnix()) {
    execute("exec-62");
  }
}

</code></pre>

<br>
<h4 style="margin:0px">Class: org.apache.commons.exec.issues.Exec65Test </h4><pre class="type-13 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * Please note that this tests make assumptions about the environment. It assumes
 * that user "root" exists and that the current user is not a "sudoer" already
 * (thereby requiring a password).
 */
@Test(expected=ExecuteException.class,timeout=TEST_TIMEOUT) public void testExec65WithSudoUsingShellScript() throws Exception {
  if (OS.isFamilyUnix()) {
    final DefaultExecutor executor=new DefaultExecutor();
    executor.setStreamHandler(new PumpStreamHandler(System.out,System.err,System.in));
    executor.setWatchdog(new ExecuteWatchdog(WATCHDOG_TIMEOUT));
    final CommandLine command=new CommandLine(resolveTestScript("issues","exec-65"));
    executor.execute(command);
  }
 else {
    throw new ExecuteException(testNotSupportedForCurrentOperatingSystem(),0);
  }
}

</code></pre>

<br>
<pre class="type-13 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Executes methods or other tests from the same test unit
"></span><br>
@Test(expected=ExecuteException.class,timeout=TEST_TIMEOUT) public void testExec65WitSleepUsingSleepCommandDirectly() throws Exception {
  if (OS.isFamilyUnix()) {
    final ExecuteWatchdog watchdog=new ExecuteWatchdog(WATCHDOG_TIMEOUT);
    final DefaultExecutor executor=new DefaultExecutor();
    final CommandLine command=new CommandLine("sleep");
    command.addArgument("60");
    executor.setStreamHandler(new PumpStreamHandler(System.out,System.err));
    executor.setWatchdog(watchdog);
    executor.execute(command);
  }
 else {
    throw new ExecuteException(testNotSupportedForCurrentOperatingSystem(),0);
  }
}

</code></pre>

<br>
<pre class="type-13 type-9 "><code><span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Verifies that exceptions are thrown during the test case execution">ExceptionVerifier</span>&nbsp;<span class="label label-info" style="display: inline-block;" data-toggle="tooltip" title="Executes methods or other tests from the same test unit">ExecutionTester</span>&nbsp;<span class=" glyphicon glyphicon-comment" aria-hidden="true" label-info" data-toggle="tooltip" title="This method/test case: 
- Verifies that exceptions are thrown during the test case execution
- Executes methods or other tests from the same test unit
"></span><br>
/** 
 * This test currently only works for Mac OS X
 * <ul>
 * <li>Linux hangs on the process stream while the process is finished</li>
 * <li>Windows seems to have similar problems</li>
 * </ul>
 * @TODO Fix tests for Windows & Linux
 */
@Test(expected=ExecuteException.class,timeout=TEST_TIMEOUT) public void testExec65WithSleepUsingShellScript() throws Exception {
  if (OS.isFamilyMac()) {
    final DefaultExecutor executor=new DefaultExecutor();
    executor.setStreamHandler(new PumpStreamHandler(System.out,System.err));
    executor.setWatchdog(new ExecuteWatchdog(WATCHDOG_TIMEOUT));
    final CommandLine command=new CommandLine(resolveTestScript("sleep"));
    executor.execute(command);
  }
 else {
    throw new ExecuteException(testNotSupportedForCurrentOperatingSystem(),0);
  }
}

</code></pre>

<br>
<script>$(document).ready(function() {
  $('pre code').each(function(i, block) {
    hljs.highlightBlock(block);
  });
});</script>
